crates/typst-cli/src/update.rs

```diff
@@ -0,0 +1,311 @@
+            .read_to_end(&mut buffer)
+            .map_err(|err| eco_format!("failed to read binary data: {}", err))?;
+
+        Ok(buffer)
+    }
+}
+
+impl<R: Read> Unpack for tar::Archive<R> {
+    fn unpack_typst_binary(&mut self) -> StrResult<Vec<u8>> {
+        let mut binary = self
+            .entries()
+            .map_err(|err| eco_format!("xz archive corrupted: {}", err))?
+            .filter_map(Result::ok)
+            .find(|e| e.path().unwrap_or_default().ends_with("typst"))
+            .ok_or("asset archive did not contain typst binary")?;
+
+        let mut buffer = Vec::with_capacity(binary.size() as usize);
+        binary
+            .read_to_end(&mut buffer)
+            .map_err(|err| eco_format!("failed to read binary data: {}", err))?;
+
+        Ok(buffer)
+    }
+}
+
+/// The raw binary data from a packed [`Asset`].
+struct PackedBinary(Vec<u8>);
+
+impl PackedBinary {
+    /// Unpacks the asset archive in-memory and writes the uncompressed contents
+    /// into a buffer.
+    ///
+    /// In-memory refers to that the physical archive is never written to disk,
+    /// only the Typst CLI binary will be plucked from the archive and written
+    /// to disk at a later stage, the rest of the archive is discarded.
+    pub fn unpack(self) -> StrResult<Vec<u8>> {
+        let mut raw = BufReader::new(Cursor::new(self.0));
+
+        tar::Archive::new(XzDecoder::new(raw.by_ref()))
+            .unpack_typst_binary()
+            .ok()
+            .or_else(|| {
+                ZipArchive::new(raw)
+                    .ok()
+                    .and_then(|mut archive| archive.unpack_typst_binary().ok())
+            })
+            .ok_or("asset archive unknown or corrupted".into())
+    }
+}
+
+/// Determines what asset to download according to the target platform the CLI
+/// is running on.
+fn asset_needed() -> StrResult<&'static str> {
+    Ok(match env!("TARGET") {
+        "x86_64-unknown-linux-gnu" => "typst-x86_64-unknown-linux-musl",
+        "x86_64-unknown-linux-musl" => "typst-x86_64-unknown-linux-musl",
+        "aarch64-unknown-linux-musl" => "typst-aarch64-unknown-linux-musl",
+        "aarch64-unknown-linux-gnu" => "typst-aarch64-unknown-linux-musl",
+        "armv7-unknown-linux-musleabi" => "typst-armv7-unknown-linux-musleabi",
+        "x86_64-apple-darwin" => "typst-x86_64-apple-darwin",
+        "aarch64-apple-darwin" => "typst-aarch64-apple-darwin",
+        "x86_64-pc-windows-msvc" => "typst-x86_64-pc-windows-msvc",
+        target => bail!("unsupported target: {target}"),
+    })
+}
+
+/// Compares latest release version to current version to see if an update
+/// is even needed
+fn update_needed(release: &Release) -> bool {
+    let current_tag: Version = env!("CARGO_PKG_VERSION").parse().unwrap();
+    let new_tag: Version = release
+        .tag_name
+        .strip_prefix('v')
+        .unwrap_or(&release.tag_name)
+        .parse()
+        .expect("release tag not in semver format");
```

Comment:
this shouldn't panic. only programmer errors should panic, but this depends on external data to be right.

Reply:
Now uses `map_err(|_| "release tag not in semver format")?`. I could also still include the `err` here if that would be useful?

Reply:
it's fine as is